### ISSUE TITLE
Performance issues using ORDER BY and LIMIT with large OFFSETs on wide tables

### ISSUE THREAD

**Issue by swang1001** [duckdb/duckdb#11261]
### What happens?

When sorting a table with `LIMIT` `n` `OFFSET` `m`, the Top-N optimization holds `n+m` full rows in memory even though only `n` rows are required in the result. With small `n` and large `m`, this is inefficient and can run out of memory on sufficiently large (both wide and long) tables.

### To Reproduce

Construct a large table of random data (100 columns, 10M rows, ~8GB):

```python
import numpy as np
import pyarrow as pa
import pyarrow.parquet as pq

nrows = 10_000_000
data = {"id": pa.array(range(nrows))}
data |= {f"col{i}": pa.array(np.random.rand(nrows)) for i in range(99)}
tbl = pa.Table.from_pydict(data)
pq.write_table(tbl, 'test.parquet')
```

Selecting the 9 millionth row ordered by the first column runs out of memory (with a limit of 8GB and one thread):

```python
import duckdb
con = duckdb.connect()
con.execute("PRAGMA threads=1")
con.execute("PRAGMA memory_limit='8GB'")
con.execute("PRAGMA temp_directory='/tmp'")
con.execute("SELECT * FROM 'test.parquet' ORDER BY col0 LIMIT 1 OFFSET 9000000")
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
duckdb.duckdb.OutOfMemoryException: Out of Memory Error: could not allocate block of size 7.5 GiB (452.1 MiB/7.4 GiB used)
```

Manually rewriting the query to select a unique key while sorting and joining with the original table has no memory issues:

```python
con.execute("SELECT * FROM 'test.parquet' INNER JOIN (SELECT id FROM 'test.parquet' ORDER BY col0 LIMIT 1 OFFSET 9000000) USING (id)")
con.fetchall()
```
```
[(6772265, 0.8999772369092386, ...
```

If I disable the memory limit on the original query, it ends up taking 47 seconds, while the optimized form takes under 10s. But this sort of optimization is difficult to apply in a general case if our parquet doesn't have an obvious primary key to join on.

### OS:

Linux x86_64

### DuckDB Version:

0.10.1

### DuckDB Client:

Python

### Full Name:

Simin Wang

### Affiliation:

Clarity

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by szarnyasg** [comment]
Thanks @swang1001, I could reproduce this. We'll take a look.

**Comment by szarnyasg** [comment]
Hi @swang1001, we discussed this internally and this is currently an expected restriction – optimizing this is non-trivial and would likely have a negative effect on the performance of other workloads.

**Comment by ASaltedFishBoy** [comment]
I also encountered this problem, you can split the order by clause and the limit clause into two select statements to optimize performance somewhat, but in practice this still does not meet my needs

**Comment by ASaltedFishBoy** [comment]
> Hi @swang1001, we discussed this internally and this is currently an expected restriction – optimizing this is non-trivial and would likely have a negative effect on the performance of other workloads.

I am curious about the problem under this use case, can you tell me the reason in more detail? If so, I may consider modifying the relevant code separately in my own local. In addition, you can consider requiring explicit declarations by users, such as indexing, to avoid general negative performance impacts.

**Comment by xuke-hat** [comment]
> tion templates related to c++ might have design problems, such as extra memory object copies, content that can be checked at

I guess the problem here is early materialization. limit + offset rows must be kept since it is possible for any of these rows to be in the result. However not every column needs to be sorted, since only columns in `order by` are necessary. One way to solve this problem is late materialization, and it is non-trivial. I'm not sure but it seems that DuckDB uses early materialization for most operators.

**Comment by msf** [comment]
At DuneAnalytics, we're using duckdb for data-API requests. We can also confirm this perf problem.
Let us know if any more datapoints or examples are useful to solve this

**Comment by msf** [comment]
simple example: 
```
D create table foo as select * as id from generate_series(1, 100000000) as id;
D .timer on
D select * from foo where id % 10 = 1 limit 10 offset 9000;
┌───────┐
│  id   │
│ int64 │
├───────┤
│ 90001 │
│ 90011 │
│ 90021 │
│ 90031 │
│ 90041 │
│ 90051 │
│ 90061 │
│ 90071 │
│ 90081 │
│ 90091 │
└───────┘
Run Time (s): real 0.001 user 0.004192 sys 0.000390
D select * from foo where id % 10 = 1 limit 10 offset 9000000;
┌──────────┐
│    id    │
│  int64   │
├──────────┤
│ 90000001 │
│ 90000011 │
│ 90000021 │
│ 90000031 │
│ 90000041 │
│ 90000051 │
│ 90000061 │
│ 90000071 │
│ 90000081 │
│ 90000091 │
├──────────┤
│ 10 rows  │
└──────────┘
Run Time (s): real 0.258 user 0.251286 sys 0.000233

```

**Comment by d33bs** [comment]
Just adding a bump to mention a fix for this would be super helpful!

**Comment by msf** [comment]
I'm about to mention this issue on my talk at "DuckCon #5", in our production use case the performance hit is much grater than ±250millis, which I presume is due to the tables being real world tables with multiple columns and not abnormally compressible data.
Also, we observe this in native tables and in views that map to `select * from _parquet_file`

**Comment by d33bs** [comment]
@msf best wishes on your upcoming talk at DuckCon 5!

Just in case, wanted to share along a workaround which @Teggy very helpfully mentioned through the DuckDB Discord. If using these queries for pagination (or possibly even if they aren't) it could be they would benefit from keyset pagination instead of offset pagination. See here for more: https://use-the-index-luke.com/no-offset . Offset queries apparently must order, fetch, then drop the data prior to the offset for returning results (meaning large offsets could be resource intensive).  I'm unsure how this works with DuckDB in specific so it could be worth verifying about improvements still through this issue.

**Comment by msf** [comment]
> @msf best wishes on your upcoming talk at DuckCon 5!
> 
> Just in case, wanted to share along a workaround which @Teggy very helpfully mentioned through the DuckDB Discord. If using these queries for pagination (or possibly even if they aren't) it could be they would benefit from keyset pagination instead of offset pagination. See here for more: https://use-the-index-luke.com/no-offset . Offset queries apparently must order, fetch, then drop the data prior to the offset for returning results (meaning large offsets could be resource intensive). I'm unsure how this works with DuckDB in specific so it could be worth verifying about improvements still through this issue.

Yes, that approach would be the other intuitive way of doing it.. will investigate how to use the row-number from the parquet as a "Key" for the offset... Might work for us in some cases.

**Comment by jgrg** [comment]
We've recently implemented this keyset pagination, in PostgreSQL in our case. It's also known as cursor paging and can be used wherever pagination is sorted by a unique key. You store the key value (the cursor) of the last row on a page and query for `WHERE colname > 'cursor' ORDER BY colname LIMIT 100`

**Comment by carlopi** [comment]
I believe this has been solved via row_id pushdown.

I just tried to reproduce, and got:
```
>>> import duckdb
>>> duckdb.__version__
'1.4.0'
>>> con = duckdb.connect()
>>> con.execute("PRAGMA threads=1")
<_duckdb.DuckDBPyConnection object at 0x10091b2b0>
>>> con.execute("PRAGMA memory_limit='8GB'")
<_duckdb.DuckDBPyConnection object at 0x10091b2b0>
>>> con.execute("PRAGMA temp_directory='/tmp'")
<_duckdb.DuckDBPyConnection object at 0x10091b2b0>
>>> con.sql("SELECT * FROM 'test.parquet' ORDER BY col0 LIMIT 1 OFFSET 9000000").show()
100% ▕██████████████████████████████████████▏ (00:00:02.30 elapsed)
┌─────────┬────────────────────┬─────────────────────┬─────────────────────┬────────────────────┬───┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬─────────────────────┐
│   id    │        col0        │        col1         │        col2         │        col3        │ … │       col94        │       col95        │       col96        │       col97        │        col98        │
│  int64  │       double       │       double        │       double        │       double       │   │       double       │       double       │       double       │       double       │       double        │
├─────────┼────────────────────┼─────────────────────┼─────────────────────┼────────────────────┼───┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 2313341 │ 0.9001303733550448 │ 0.36232720285018005 │ 0.24217911683079996 │ 0.3000664822448663 │ … │ 0.7991552127237243 │ 0.9160682183993734 │ 0.6156468522968795 │ 0.5716831643800695 │ 0.22825760566548425 │
├─────────┴────────────────────┴─────────────────────┴─────────────────────┴────────────────────┴───┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴─────────────────────┤
│ 1 rows                                                                                                                                                                               100 columns (10 shown) │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
     
```
